File: frontend/src/api/partnerships.ts
import { fetchWithAuth } from "./client";
import type {
  KooperationStatus,
  Partnership,
  PartnershipCreate,
  PartnershipUpdate,
} from "./types";

interface ListFilters {
  search?: string;
  land?: string;
  kooperation?: string;
}

export async function getPartnerships(
  filters: ListFilters = {},
): Promise<Partnership[]> {
  const params = new URLSearchParams();
  if (filters.search) params.set("search", filters.search);
  if (filters.land) params.set("land", filters.land);
  if (filters.kooperation) params.set("kooperation", filters.kooperation);

  const query = params.toString();
  const path = query ? `/api/partnerships?${query}` : "/api/partnerships";
  return fetchWithAuth<Partnership[]>(path);
}

export async function getPartnership(id: number): Promise<Partnership> {
  return fetchWithAuth<Partnership>(`/api/partnerships/${id}`);
}

export async function createPartnership(
  data: PartnershipCreate,
): Promise<Partnership> {
  // Convert camelCase frontend keys to snake_case backend keys
  return fetchWithAuth<Partnership>("/api/partnerships", {
    method: "POST",
    body: JSON.stringify({
      park_name: data.parkName,
      land: data.land,
      stadt: data.stadt,
      gruendungsjahr: data.gruendungsjahr,
      bisherige_kooperation: data.bisherigeKooperation,
      datum: data.datum,
      themen: data.themen,
      bemerkungen: data.bemerkungen,
      park_ansprechpartner: data.parkAnsprechpartner,
      kontaktdetails: data.kontaktdetails,
      webpraesenz: data.webpraesenz,
      universitaet_name: data.universitaetName,
      standort: data.standort,
      forschungsschwerpunkte: data.forschungsschwerpunkte,
      uni_ansprechpartner: data.uniAnsprechpartner,
      website: data.website,
    }),
  });
}

export async function updatePartnership(
  id: number,
  data: PartnershipUpdate,
): Promise<Partnership> {
  // Only send fields that are explicitly set
  const body: Record<string, unknown> = {};
  if (data.parkName !== undefined) body.park_name = data.parkName;
  if (data.land !== undefined) body.land = data.land;
  if (data.stadt !== undefined) body.stadt = data.stadt;
  if (data.gruendungsjahr !== undefined) body.gruendungsjahr = data.gruendungsjahr;
  if (data.bisherigeKooperation !== undefined) body.bisherige_kooperation = data.bisherigeKooperation;
  if (data.datum !== undefined) body.datum = data.datum;
  if (data.themen !== undefined) body.themen = data.themen;
  if (data.bemerkungen !== undefined) body.bemerkungen = data.bemerkungen;
  if (data.parkAnsprechpartner !== undefined) body.park_ansprechpartner = data.parkAnsprechpartner;
  if (data.kontaktdetails !== undefined) body.kontaktdetails = data.kontaktdetails;
  if (data.webpraesenz !== undefined) body.webpraesenz = data.webpraesenz;
  if (data.universitaetName !== undefined) body.universitaet_name = data.universitaetName;
  if (data.standort !== undefined) body.standort = data.standort;
  if (data.forschungsschwerpunkte !== undefined) body.forschungsschwerpunkte = data.forschungsschwerpunkte;
  if (data.uniAnsprechpartner !== undefined) body.uni_ansprechpartner = data.uniAnsprechpartner;
  if (data.website !== undefined) body.website = data.website;

  return fetchWithAuth<Partnership>(`/api/partnerships/${id}`, {
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export async function updateKooperationStatus(
  id: number,
  status: KooperationStatus,
): Promise<Partnership> {
  return fetchWithAuth<Partnership>(`/api/partnerships/${id}/status`, {
    method: "PATCH",
    body: JSON.stringify({ bisherige_kooperation: status }),
  });
}

export async function deletePartnership(id: number): Promise<void> {
  return fetchWithAuth<void>(`/api/partnerships/${id}`, {
    method: "DELETE",
  });
}

File: backend/app/routers/partnerships.py
"""Partnership API endpoints."""

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.auth.dependencies import get_current_user, require_role
from app.crud import partnership as crud
from app.database import get_db
from app.models import SciencePark, User
from app.schemas.partnership import (
    KooperationStatusUpdate,
    PartnershipCreate,
    PartnershipResponse,
    PartnershipUpdate,
)

router = APIRouter(prefix="/api/partnerships", tags=["partnerships"])


def _to_response(park: SciencePark) -> PartnershipResponse:
    uni = park.university
    return PartnershipResponse(
        id=park.id,
        park_name=park.name,
        land=park.land,
        stadt=park.stadt,
        gruendungsjahr=park.gruendungsjahr,
        bisherige_kooperation=park.bisherige_kooperation,
        datum=park.datum,
        themen=park.themen or [],
        bemerkungen=park.bemerkungen,
        park_ansprechpartner=park.ansprechpartner,
        kontaktdetails=park.kontaktdetails,
        webpraesenz=park.webpraesenz,
        universitaet_name=uni.name if uni else "",
        standort=uni.standort if uni else None,
        forschungsschwerpunkte=uni.forschungsschwerpunkte or [] if uni else [],
        uni_ansprechpartner=uni.ansprechpartner if uni else None,
        website=uni.website if uni else None,
        created_at=park.created_at,
        updated_at=park.updated_at,
    )


@router.get("", response_model=list[PartnershipResponse])
def list_partnerships(
    search: str | None = Query(default=None),
    land: str | None = Query(default=None),
    kooperation: str | None = Query(default=None),
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
) -> list[PartnershipResponse]:
    parks = crud.get_partnerships(db, search=search, land=land, kooperation=kooperation)
    return [_to_response(p) for p in parks]


@router.get("/{partnership_id}", response_model=PartnershipResponse)
def get_partnership(
    partnership_id: int,
    db: Session = Depends(get_db),
    _user: User = Depends(get_current_user),
) -> PartnershipResponse:
    park = crud.get_partnership(db, partnership_id)
    if not park:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Partnerschaft nicht gefunden.",
        )
    return _to_response(park)


@router.post(
    "", response_model=PartnershipResponse, status_code=status.HTTP_201_CREATED
)
def create_partnership(
    data: PartnershipCreate,
    db: Session = Depends(get_db),
    _user: User = Depends(require_role("readwrite")),
) -> PartnershipResponse:
    park = crud.create_partnership(db, data)
    return _to_response(park)


@router.put("/{partnership_id}", response_model=PartnershipResponse)
def update_partnership(
    partnership_id: int,
    data: PartnershipUpdate,
    db: Session = Depends(get_db),
    _user: User = Depends(require_role("readwrite")),
) -> PartnershipResponse:
    park = crud.update_partnership(db, partnership_id, data)
    if not park:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Partnerschaft nicht gefunden.",
        )
    return _to_response(park)


@router.patch(
    "/{partnership_id}/status", response_model=PartnershipResponse
)
def update_status(
    partnership_id: int,
    data: KooperationStatusUpdate,
    db: Session = Depends(get_db),
    _user: User = Depends(require_role("readwrite")),
) -> PartnershipResponse:
    park = crud.update_status(db, partnership_id, data.bisherige_kooperation)
    if not park:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Partnerschaft nicht gefunden.",
        )
    return _to_response(park)


@router.delete(
    "/{partnership_id}", status_code=status.HTTP_204_NO_CONTENT
)
def delete_partnership(
    partnership_id: int,
    db: Session = Depends(get_db),
    _user: User = Depends(require_role("admin")),
) -> None:
    deleted = crud.delete_partnership(db, partnership_id)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Partnerschaft nicht gefunden.",
        )
